csrc/optimization/consecutive_cast.cpp

```diff
@@ -0,0 +1,195 @@
+// dtype in the chain. i.e. think about pairs like fp16/bfloat16, or fp32/int32,
+// where one can't represent the other. In order to handle this scenario, we can
+// just keep track of a `starting_anchor` that indicates the starting point of
+// the section, that has a valid `lo_anchor`. When we encounter a new cast op
+// that breaks the assumption, we'll optimize what we have seen in the existing
+// section and start a new section with the next cast op.
+//
+// The algorithm:
+// 1. iterating through all expr in the fusion:
+//    1.1 we skip all exprs other than cast;
+//    1.2 for each end cast-op 'expr', we trace back its producers iteratively
+//    and push the value(s) on top of `chain_cast_vals`, until:
+//
+//        a. the producer is not a cast op; or
+//
+//        b. the producer is used by other ops, or is a fusion output.
+//
+//    1.3 at this point, each `chain_cast_vals` has an ordered cast outputs with
+//    a straight line dependency:
+//        1.3.1 we point starting_anchor at the beginning op, indicating the
+//        starting point of our folding optimization, meanwhile, we point
+//        lo_anchor at the first op, indicating the narrowest dtype we have seen
+//        in the segment;
+//        1.3.2 we enter the loop to iterate through items
+//        inside `chain_cast_vals`, for item `val`:
+//
+//              a. if `val_dtype` is the same as, or wider than `anchor_dtype`
+//              of `lo_anchor`, current cast is a no-op and can be ignored;
+//
+//              b. if `anchor_dtype` is narrower than `val_dtype`, previous cast
+//              to `lo_anchor` is a no-op and can be folded away. We update
+//              `lo_anchor` to point to `val`;
+//
+//              c. otherwise, `val` and `lo_anchor` are incompatible casts and
+//              both needs to be preserved. We'll rewire it as:
+//              `starting_anchor`->`lo_anchor`->`val`. Afterwards, we'll update
+//              `starting_anchor` and `lo_anchor` to both point at `val`.
+//
+//    1.4 At this point we look at `anchor_dtype` of `lo_anchor` and
+//    `output_dtype` of `expr->output(0)`:
+//
+//        a. if `anchor_dtype` is the same as `output_dtype`, we skip the last
+//        cast op and replace all its uses with `lo_anchor`;
+//
+//        b. if `anchor_dtype` is wider than `output_dtype`, all previous cast
+//        after `starting_anchor` is no-op, we re-wire `starting_anchor`
+//        directly to `expr`;
+//
+//        c. otherwise, we can't bypass `lo_anchor` cast, we rewire this
+//        section as `starting_anchor`->`lo_anchor`->`expr->output(0)`
+void castOptimizationPass(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  std::unordered_set<Expr*> visited;
+  for (int i = exprs.size() - 1; i >= 0; --i) {
+    auto expr = exprs[i];
+    // skip current expr if it's not a foldable cast or it has already been
+    // addressed
+    if (visited.count(expr) != 0 || !isCast(expr)) {
```

Comment:
nit:
```
```suggestion
    if (!isCast(expr) || visited.count(expr) != 0) {
```
Because I think `isCast` is faster to check and more likely to happen, so we should put it first to short-circuit the later relatively expensive check.

Reply:
visited check should be dirt cheap though right?!

Also, unfortunately I was getting a segfault earlier with that.
It's very surprising to me that apparently some fusion mutation somehow makes `expr` in a limbo state. i.e. without the short-cut of `visited.count(expr)`, The dynamic cast on `expr` gave me a segfault... which is pretty surprising.....

Reply:
FYI, I think this is a desired behavior. i.e. for expr that has been short-wired, it's being removed from the graph and that memory is also reused for something else?!

I'm just surprised that we already have that. I thought our mutation passes are only *adding* exprs...

Reply:
> I thought our mutation passes are only adding exprs...
Haha, I am thinking  the same.

But whatever, they are both super cheap. It's just cheap vs cheaper.